nbgrader/api.py

```diff
@@ -1,618 +1,1169 @@
+
+    id = Column(String(32), primary_key=True, default=new_uuid)
+    cell_id = Column(String(32), ForeignKey('grade_cell.id'))
+    notebook_id = Column(String(32), ForeignKey('submitted_notebook.id'))
+
+    auto_score = Column(Float())
+    manual_score = Column(Float())
+
+    assignment = association_proxy('notebook', 'assignment')
+    student = association_proxy('notebook', 'student')
+    max_score = association_proxy('cell', 'max_score')
+
+    @property
+    def score(self):
+        if self.manual_score is not None:
+            return self.manual_score
+        elif self.auto_score is not None:
+            return self.auto_score
+        else:
+            return 0.0
+
+    @property
+    def needs_manual_grade(self):
+        return self.cell.cell_type == "markdown" and self.manual_score is None
+
+    def to_dict(self):
+        return {
+            "id": self.id,
+            "name": self.cell.name,
+            "notebook": self.notebook.notebook.name,
+            "assignment": self.assignment.assignment.name,
+            "student": self.student.id,
+            "auto_score": self.auto_score,
+            "manual_score": self.manual_score,
+            "max_score": self.max_score,
+            "needs_manual_grade": self.needs_manual_grade
+        }
+
+    def __repr__(self):
+        return "{} for {}".format(self.cell, self.student)
+
+
+class Comment(Base):
+    __tablename__ = "comment"
+    __table_args__ = (UniqueConstraint('cell_id', 'notebook_id'),)
+
+    id = Column(String(32), primary_key=True, default=new_uuid)
+    cell_id = Column(String(32), ForeignKey('solution_cell.id'))
+    notebook_id = Column(String(32), ForeignKey('submitted_notebook.id'))
+
+    comment = Column(Text())
+
+    assignment = association_proxy('notebook', 'assignment')
+    student = association_proxy('notebook', 'student')
+
+    def to_dict(self):
+        return {
+            "id": self.id,
+            "name": self.cell.name,
+            "notebook": self.notebook.notebook.name,
+            "assignment": self.assignment.assignment.name,
+            "student": self.student.id,
+            "comment": self.comment,
+        }
+
+    def __repr__(self):
+        return "{} for {}".format(self.cell, self.student)
+
+
+class Gradebook(object):
+    """The gradebook object to interface with the database holding
+    nbgrader grades.
+
+    """
+
+    def __init__(self, db_url):
+        """Initialize the connection to the database.
+
+        Parameters
+        ----------
+        db_url : string
+            The URL to the database, e.g. sqlite:///grades.db
 
         """
-        return self._find_all('grade_cells', attributes)
+        # create the connection to the database
+        engine = create_engine(db_url)
+        self.db = scoped_session(sessionmaker(autoflush=True, bind=engine))
+        Base.query = self.db.query_property()
+        Base.metadata.create_all(bind=engine)
```

Comment:
If I'm right about that, maybe there's a better way to abstract this.  Minimally we could throw an exception when the user tries to construct another instance without disposing of the first.